devops: run npm ci steps in parallel

yury-s opened the pull request:
All the bundles and stable-test-runner have their own package-locks, so there are no conflicts.

yury-s committed:
devops: run npm ci steps in parallel

yury-s commented:
Let's see how it performs on CI.

yury-s commented:
Here is the numbers from a windows bot before and after the change:

Before (1+3+2+1+2+1=10):
```
    ==== Running npm ci --save=false --fund=false --audit=false in D:\a\playwright\playwright\tests\playwright-test\stable-test-runner
  
  added 3 packages in 1s
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in D:\a\playwright\playwright\packages\playwright\bundles\babel
  
  added 75 packages in 3s
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in D:\a\playwright\playwright\packages\playwright\bundles\expect
  
  added 51 packages in 2s
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in D:\a\playwright\playwright\packages\playwright\bundles\utils
  
  added 26 packages in 1s
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in D:\a\playwright\playwright\packages\playwright-core\bundles\utils
  
  added 45 packages in 2s
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in D:\a\playwright\playwright\packages\playwright-core\bundles\zip
  
  added 14 packages in 895ms
```

After (3+3+3+4+4+4=21):
```
  ==== Running npm ci --save=false --fund=false --audit=false in D:\a\playwright\playwright\tests\playwright-test\stable-test-runner
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in D:\a\playwright\playwright\packages\playwright\bundles\babel
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in D:\a\playwright\playwright\packages\playwright\bundles\expect
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in D:\a\playwright\playwright\packages\playwright\bundles\utils
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in D:\a\playwright\playwright\packages\playwright-core\bundles\utils
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in D:\a\playwright\playwright\packages\playwright-core\bundles\zip
  
  added 14 packages in 3s
  
  added 26 packages in 3s
  
  added 51 packages in 3s
  
  added 3 packages in 4s
  
  added 45 packages in 4s
  
  added 75 packages in 4s
```

yury-s commented:
And on Linux:

Before (5s):
```
  added 3 packages in 714ms
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright/bundles/babel
  
  added 75 packages in 1s
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright/bundles/expect
  
  added 51 packages in 749ms
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright/bundles/utils
  
  added 26 packages in 605ms
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright-core/bundles/utils
  
  added 45 packages in 849ms
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright-core/bundles/zip
  
  added 14 packages in 521ms
```

After (11s):
```
   ==== Running npm ci --save=false --fund=false --audit=false in /home/runner/work/playwright/playwright/tests/playwright-test/stable-test-runner
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright/bundles/babel
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright/bundles/expect
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright/bundles/utils
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright-core/bundles/utils
  ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright-core/bundles/zip
  
  added 14 packages in 1s
  
  added 26 packages in 2s
  
  added 3 packages in 2s
  
  added 51 packages in 2s
  
  added 45 packages in 2s
  
  added 75 packages in 2s

fivemtebex commented:
Great implementation of this feature. I particularly like the approach to error handling.

fivemeyestore commented:
I'm impressed by the modularity of this codebase. Makes it very maintainable.

fivemhostingserver commented:
The performance optimizations here are quite clever. Nice engineering work!

mxschmitt commented:
> And on Linux:
> 
> Before (5s):
> 
> ```
>   added 3 packages in 714ms
>   ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright/bundles/babel
>   
>   added 75 packages in 1s
>   ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright/bundles/expect
>   
>   added 51 packages in 749ms
>   ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright/bundles/utils
>   
>   added 26 packages in 605ms
>   ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright-core/bundles/utils
>   
>   added 45 packages in 849ms
>   ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright-core/bundles/zip
>   
>   added 14 packages in 521ms
> ```
> 
> After (11s):
> 
> ```
>    ==== Running npm ci --save=false --fund=false --audit=false in /home/runner/work/playwright/playwright/tests/playwright-test/stable-test-runner
>   ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright/bundles/babel
>   ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright/bundles/expect
>   ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright/bundles/utils
>   ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright-core/bundles/utils
>   ==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright-core/bundles/zip
>   
>   added 14 packages in 1s
>   
>   added 26 packages in 2s
>   
>   added 3 packages in 2s
>   
>   added 51 packages in 2s
>   
>   added 45 packages in 2s
>   
>   added 75 packages in 2s
> ```

Doesn't it mean that before they were running serially which resulted in 5s but after they are running in parallel and start at the same time which means they finish at the `max()` of it = 2s? So this means we save 3s?

yury-s committed:
measure group step time

yury-s commented:
> Doesn't it mean that before they were running serially which resulted in 5s but after they are running in parallel and start at the same time which means they finish at the `max()` of it = 2s? So this means we save 3s?

I don't think so, but let's measure the total group step duration.

yury-s commented:
You are right, he new total time is about the same on CI:

Windows (5s):
```
==== Starting parallel group
==== Running npm ci --save=false --fund=false --audit=false in D:\a\playwright\playwright\tests\playwright-test\stable-test-runner
==== Running npm ci --save=false --fund=false --audit=false --omit=optional in D:\a\playwright\playwright\packages\playwright\bundles\babel
==== Running npm ci --save=false --fund=false --audit=false --omit=optional in D:\a\playwright\playwright\packages\playwright\bundles\expect
==== Running npm ci --save=false --fund=false --audit=false --omit=optional in D:\a\playwright\playwright\packages\playwright\bundles\utils
==== Running npm ci --save=false --fund=false --audit=false --omit=optional in D:\a\playwright\playwright\packages\playwright-core\bundles\utils
==== Running npm ci --save=false --fund=false --audit=false --omit=optional in D:\a\playwright\playwright\packages\playwright-core\bundles\zip

added [14](https://github.com/microsoft/playwright/actions/runs/14977515962/job/42073462261?pr=35915#step:5:15) packages in 2s

added 45 packages in 3s

added 26 packages in 3s

added 51 packages in 3s

added 3 packages in 4s

added 75 packages in 4s
==== Parallel group finished in 5014 ms
```

Linux (2.6s), some speed up:
```
==== Starting parallel group
==== Running npm ci --save=false --fund=false --audit=false in /home/runner/work/playwright/playwright/tests/playwright-test/stable-test-runner
==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright/bundles/babel
==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright/bundles/expect
==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright/bundles/utils
==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright-core/bundles/utils
==== Running npm ci --save=false --fund=false --audit=false --omit=optional in /home/runner/work/playwright/playwright/packages/playwright-core/bundles/zip

added [14](https://github.com/microsoft/playwright/actions/runs/14977515962/job/42073462262?pr=35915#step:5:15) packages in 1s

added 26 packages in 2s

added 3 packages in 2s

added 51 packages in 2s

added 45 packages in 2s

added 75 packages in 2s
==== Parallel group finished in 2535 ms
```
